jbi/queue.py

```diff
@@ -0,0 +1,315 @@
+        return list(sorted(items, key=lambda i: i.payload.event.time or 0))
+
+    async def get_all(self) -> dict[int, list[QueueItem]]:
+        all_items: dict[int, list[QueueItem]] = defaultdict(list)
+        for filepath in self.location.rglob("*.json"):
+            item = QueueItem.parse_file(filepath)
+            bisect.insort_right(
+                all_items[item.payload.bug.id],
+                item,
+                key=lambda i: i.payload.event.time or 0,
+            )
+        return all_items
+
+    @property
+    def size(self) -> int:
+        return sum(1 for _ in self.location.rglob("*.json"))
+
+
+class InvalidQueueDSNError(Exception):
+    pass
+
+
+class DeadLetterQueue:
+    backend: QueueBackend
+
+    def __init__(self, dsn: DLQUrl | str | ParseResult):
+        if isinstance(dsn, str):
+            dsn = urlparse(url=dsn)
+
+        if dsn.scheme == "memory":
+            self.backend = MemoryBackend()
+        elif dsn.scheme == "file":
+            self.backend = FileBackend(dsn.path)
+        else:
+            raise InvalidQueueDSNError(f"{dsn.scheme} is not supported")
+
+    def ready(self):
+        """Heartbeat check to assert we can write items to queue
+
+        TODO: Convert to an async method when Dockerflow's FastAPI integration
+        can run check asynchronously
+        """
+
+        ping_result = self.backend.ping()
+        if ping_result is False:
+            return [
+                checks.Error(f"queue with f{str(self.backend)} backend unavailable")
+            ]
+        return []
+
+    async def postpone(self, payload: bugzilla.WebhookRequest):
+        """
+        Postpone the specified request for later.
+        """
+        item = QueueItem(payload=payload)
+        await self.backend.put(item)
+
+    async def track_failed(self, payload: bugzilla.WebhookRequest, exc: Exception):
+        """
+        Store the specified payload and exception information into the queue.
+        """
+        item = QueueItem(
+            payload=payload,
+            error=PythonException.from_exc(exc),
+        )
+        await self.backend.put(item)
+
+    async def is_blocked(self, payload: bugzilla.WebhookRequest) -> bool:
+        """
+        Return `True` if the specified `payload` is blocked and should be
+        queued instead of being processed.
+        """
+        existing = await self.backend.get(payload.bug.id)
+        return len(existing) > 0
+
+    async def retrieve(self) -> list[QueueItem]:
+        """
+        Returns the whole list of items in the queue, grouped by bug, and
+        sorted from oldest to newest.
+        """
+        existing = await self.backend.get_all()
+        # Sort by event datetime ascending.
+        by_oldest_bug_events = sorted(
+            existing.values(), key=lambda items: items[0].payload.event.time or 0
+        )
+        return list(itertools.chain(*by_oldest_bug_events))
```

Comment:
Like I was [noting in #903](https://github.com/mozilla/jira-bugzilla-integration/pull/903#discussion_r1521286965)

After thinking about this, we shouldn't flatten this list. But if the scheduler fails to process one request, it will want to jump to the next bug. 

```python

for items in all_items:
    for item in items:
        try:
            execute_action(item.payload, actions)
        except IgnoreInvalidRequestError:
            await queue.done(item)
        except Exception as exc:
            # Jump to next bug
            break
        else:
            await queue.done(item)
```

Reply:
Yeah, I think this is the right call / something for you to take into account with the retry script. @alexcottner , what do you think?